### DISCUSSION TITLE
chainloading from plain old Etherboot

### DISCUSSION THREAD

**Discussion by perkelix** [ipxe/ipxe#681] [Q&A]
Among my collection of PC hardware, I have a few rarities whose netboot implementation predates PXE.

I tried chainloading iPXE via BOOTP options. I prepared a tagged ELF image as follow:

$ sudo mkelf-linux --param=autoboot --output=/srv/tftp/ipxe.nbi ~/Projects/ipxe/src/bin/ipxe.lkrn

The host succesufully loads the iPXE blob, except that it fails to find the network card. I'm left with a flashing cursor after the initial iPXE credits:

![IMG_20220717_210749](https://user-images.githubusercontent.com/57731134/179450342-7afcebdb-1a3f-48ed-8549-e5c613fd6a58.jpg)

Any idea on how to fix this?

**Comment by stappersg** [comment]
On Sun, Jul 17, 2022 at 10:39:51PM -0700, perkelix wrote:
> iPXE 1.21.1+ (gd3c89) -- Open Source Network Boot Firmware   ....

That is indeed the latest from git.

 
> Any idea on how to fix this?

Try `make DEBUG=init`  and see if there is anything you expect from
this:

|$ grep DBG src/core/init.c
|			DBGC ( colour, "INIT startup %s...\n",
|	DBGC ( colour, "INIT startup complete\n" );
|			DBGC ( colour, "INIT shutdown %s...\n",
|	DBGC ( colour, "INIT shutdown complete\n" );
|$


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Reply by perkelix** [reply]
![IMG_20220718_091834](https://user-images.githubusercontent.com/57731134/179455053-8bf3c62f-63fc-4466-8718-ec8cd454fa41.jpg)

**Reply by stappersg** [reply]
I do read that as _it didn't dies in `init.c`_

```c
$ tail -n 19 core/main.c
__asmcall int main ( void ) {
	int rc;

	/* Perform one-time-only initialisation (e.g. heap) */
	initialise();

	/* Some devices take an unreasonably long time to initialise */
	printf ( "%s initialising devices...", product_short_name );
	startup();
	printf ( "ok\n" );

	/* Attempt to boot */
	if ( ( rc = ipxe ( NULL ) ) != 0 )
		goto err_ipxe;

 err_ipxe:
	shutdown_exit();
	return rc;
}
```

I don't know which `DEBUG=....` to use next.
And do hope that an advice is given.

**Comment by NiKiZe** [comment]
What hardware is this on? (Which CPU)
And have you tried gPXE? Use git checkout 2a36703

I would start out with same era version of the software and then bisect from there.
Is there any documentation on how to chain Etherboot from Etherboot? I think that should still apply to both gPXE and iPXE.

**Reply by stappersg** [reply]
> and then bisect from there.

Find at https://ipxe.org/howto/bisect#the_bisection_process information how to do that.

**Reply by perkelix** [reply]
This is a Geode SC2200 with a NatSemi NIC.

gPXE won't compile using the current toolchain.

**Reply by NiKiZe** [reply]
Then you will have to get a different toolchain ;)

**Comment by perkelix** [comment]
Tried a slightly different approach:

$ sudo mknbi-rom --output=/srv/tftp/ipxe.0 ~/Projects/ipxe/src/bin/natsemi.mrom

![IMG_20220718_101004](https://user-images.githubusercontent.com/57731134/179460748-aa3746fd-bf41-4e3e-b0a1-109379a09597.jpg)

Slightly different output, but still doesn't proceed beyond loading the iPXE blog.

**Reply by stappersg** [reply]
> but still doesn't proceed beyond printing some early banner.

_add additional (debug) print statements_

**Comment by perkelix** [comment]
$ sudo mknbi-rom --output=/srv/tftp/ipxe.0 ~/Projects/ipxe/src/bin/natsemi.kkpxe

![IMG_20220718_101824](https://user-images.githubusercontent.com/57731134/179461846-a8541201-8cff-4c39-a7d6-98d7e406caa0.jpg)

**Reply by NiKiZe** [reply]
kkpxe will not work, you are trying to put in a nail with a screwdriver.

**Comment by perkelix** [comment]
Apparently, the NBI target still exists in Makefile, but it's not documented on the wiki. No need for mknbi. Nice. Let's try.

$ sudo cp ~/Projects/ipxe/src/bin/natsemi.nbi /srv/tftp/ipxe.0

![IMG_20220718_120013](https://user-images.githubusercontent.com/57731134/179478394-1139ffdc-e08e-4548-9193-d3c85c2ab1c9.jpg)

Nope.

**Comment by perkelix** [comment]
Someone pointed out that this particular device might have its NIC on the ISA bus. I tried:

$ sudo mknbi-rom --output=/srv/tftp/ipxe.0 ~/Projects/ipxe/src/bin/natsemi.isarom

![IMG_20220718_132625](https://user-images.githubusercontent.com/57731134/179493080-48792110-c80b-4100-a0f8-929168a5e504.jpg)

**Reply by NiKiZe** [reply]
Well, don't assume, check 0x100b 0x0020

Valid PCI Ethernet device matching FA311 description https://pci-ids.ucw.cz/read/PC/100b/0020

**Comment by perkelix** [comment]
For comparision:

$ sudo mknbi-rom --output=/srv/tftp/ipxe.0 ~/Projects/ipxe/src/bin/natsemi.pcirom

![IMG_20220718_132830](https://user-images.githubusercontent.com/57731134/179493601-6e74b5aa-d504-4293-9440-edc7c6780bf9.jpg)

**Comment by perkelix** [comment]
I inquired and the NIC in that device is a [DP83816](https://html.alldatasheet.com/html-pdf/182817/NSC/DP83816AVNG/58/1/DP83816AVNG.html) that runs off PCI. Screenshots above report PCI ID 0x100b 0x0020 for that NIC, which is the same as for the DP83815 listed as supported by iPXE, so this should work.
